File: modules/expo-driver-location/android/src/main/java/expo/modules/driverlocation/ExpoDriverLocationModule.kt
package expo.modules.driverlocation

import android.content.Context
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoDriverLocationModule : Module() {
  private fun resolveContext(): Context? {
    return appContext.reactContext?.applicationContext
      ?: DriverLocationService.getApplicationContext()
  }

  override fun definition() = ModuleDefinition {
    Name("ExpoDriverLocation")

    Events("onShiftForceClosed")

    OnCreate {
      ShiftLocationGuardService.eventSink = { name, body ->
        try {
          sendEvent(name, body)
        } catch (_: Exception) {
          // JS bridge may be gone during app kill
        }
      }
    }

    OnDestroy {
      ShiftLocationGuardService.eventSink = null
    }

    AsyncFunction("startTracking") { config: Map<String, Any?> ->
      val context = resolveContext() ?: return@AsyncFunction null
      val trackingConfig = TrackingConfig.fromMap(config)
      TrackingSessionStore.save(context, trackingConfig)

      val action = if (DriverLocationService.isRunning) {
        DriverLocationService.ACTION_UPDATE
      } else {
        DriverLocationService.ACTION_START
      }

      try {
        DriverLocationService.launchService(context, action)
      } catch (_: Exception) {
        if (DriverLocationService.isRunning) {
          context.startService(DriverLocationService.startServiceIntent(context, action))
        }
      }
      null
    }

    AsyncFunction("stopTracking") {
      val context = resolveContext() ?: return@AsyncFunction null
      context.startService(
        DriverLocationService.startServiceIntent(context, DriverLocationService.ACTION_STOP),
      )
      null
    }

    AsyncFunction("isTracking") {
      DriverLocationService.isRunning
    }

    AsyncFunction("updateNotificationLabels") { title: String, body: String ->
      val context = resolveContext() ?: return@AsyncFunction null
      TrackingSessionStore.updateNotificationLabels(context, title, body)
      if (DriverLocationService.isRunning) {
        DriverLocationService.updateNotification(context, title, body)
      }
      null
    }

    AsyncFunction("getLastLocation") {
      val context = resolveContext() ?: return@AsyncFunction null
      val coord = TrackingSessionStore.getLastLocation(context) ?: return@AsyncFunction null
      mapOf(
        "latitude" to coord.latitude,
        "longitude" to coord.longitude,
        "heading" to (coord.heading ?: 0.0),
        "speed" to (coord.speed ?: 0.0),
        "accuracy" to (coord.accuracy ?: 0.0),
        "capturedAtMs" to (coord.capturedAtMs ?: 0.0),
      )
    }

    AsyncFunction("enableShiftLocationGuard") { config: Map<String, Any?> ->
      val context = resolveContext() ?: return@AsyncFunction null
      val trackingConfig = TrackingConfig.fromMap(config)
      val endTripApiUrl = config["endTripApiUrl"] as? String
      ShiftGuardSessionStore.save(context, trackingConfig, endTripApiUrl)

      val seedLat = (config["seedLatitude"] as? Number)?.toDouble()
      val seedLon = (config["seedLongitude"] as? Number)?.toDouble()
      if (seedLat != null && seedLon != null && seedLat != 0.0 && seedLon != 0.0) {
        ShiftGuardSessionStore.saveLastLocation(
          context,
          DriverCoordinate(
            latitude = seedLat,
            longitude = seedLon,
            heading = null,
            speed = null,
            accuracy = null,
          ),
        )
      }

      try {
        ShiftLocationGuardService.launch(context, ShiftLocationGuardService.ACTION_ENABLE)
      } catch (_: Exception) {
        context.startService(
          ShiftLocationGuardService.startServiceIntent(
            context,
            ShiftLocationGuardService.ACTION_ENABLE,
          ),
        )
      }
      null
    }

    AsyncFunction("disableShiftLocationGuard") {
      val context = resolveContext() ?: return@AsyncFunction null
      try {
        ShiftLocationGuardService.launch(context, ShiftLocationGuardService.ACTION_DISABLE)
      } catch (_: Exception) {
        context.startService(
          ShiftLocationGuardService.startServiceIntent(
            context,
            ShiftLocationGuardService.ACTION_DISABLE,
          ),
        )
      }
      ShiftGuardSessionStore.clear(context)
      null
    }

    AsyncFunction("consumePendingShiftClose") {
      val context = resolveContext() ?: return@AsyncFunction null
      ShiftGuardSessionStore.consumePendingCloseReason(context)
    }
  }
}

File: modules/expo-driver-location/ios/ExpoDriverLocationModule.swift
import ExpoModulesCore

public final class ExpoDriverLocationModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoDriverLocation")

    Events("onShiftForceClosed")

    OnCreate {
      ShiftLocationGuardManager.shared.onForceClosed = { [weak self] body in
        self?.sendEvent("onShiftForceClosed", body)
      }
    }

    OnDestroy {
      ShiftLocationGuardManager.shared.onForceClosed = nil
    }

    AsyncFunction("startTracking") { (config: [String: Any]) in
      let trackingConfig = try TrackingConfig.fromDictionary(config)
      try DriverLocationManager.shared.startTracking(config: trackingConfig)
    }

    AsyncFunction("stopTracking") {
      DriverLocationManager.shared.stopTracking(sendDeactivate: true)
    }

    AsyncFunction("isTracking") {
      DriverLocationManager.shared.isTracking
    }

    AsyncFunction("updateNotificationLabels") { (_ title: String, _ body: String) in
      DriverLocationManager.shared.updateNotificationLabels(title: title, body: body)
    }

    AsyncFunction("getLastLocation") { () -> [String: Any]? in
      guard let coord = TrackingSessionStore.getLastLocation() else {
        return nil
      }
      return [
        "latitude": coord.latitude,
        "longitude": coord.longitude,
        "heading": coord.heading as Any,
        "speed": coord.speed as Any,
        "accuracy": coord.accuracy as Any,
        "capturedAtMs": coord.capturedAtMs as Any,
      ]
    }

    // Scan → status_update: fresh GPS + replace published 15-min cache (no live-location POST).
    AsyncFunction("getFreshLocationAndPublish") { (promise: Promise) in
      ScanFreshLocationFetcher.shared.fetch { location in
        var published: DriverCoordinate? = nil
        if let location {
          published = ScanFreshLocationFetcher.shared.publishIfAcceptable(location)
        }
        // If fresh rejected, still return current published cache for status_update fallback.
        let coord = published ?? TrackingSessionStore.getLastLocation()
        guard let coord, coord.latitude != 0, coord.longitude != 0 else {
          promise.resolve(nil)
          return
        }
        promise.resolve([
          "latitude": coord.latitude,
          "longitude": coord.longitude,
          "heading": coord.heading as Any,
          "speed": coord.speed as Any,
          "accuracy": coord.accuracy as Any,
          "capturedAtMs": coord.capturedAtMs as Any,
        ])
      }
    }

    AsyncFunction("enableShiftLocationGuard") { (config: [String: Any]) in
      let trackingConfig = try TrackingConfig.fromDictionary(config)
      let endTripApiUrl = config["endTripApiUrl"] as? String
      let seedLat = (config["seedLatitude"] as? NSNumber)?.doubleValue
      let seedLon = (config["seedLongitude"] as? NSNumber)?.doubleValue
      ShiftLocationGuardManager.shared.enable(
        config: trackingConfig,
        endTripApiUrl: endTripApiUrl,
        seedLatitude: seedLat,
        seedLongitude: seedLon
      )
    }

    AsyncFunction("disableShiftLocationGuard") {
      ShiftLocationGuardManager.shared.disable()
    }

    AsyncFunction("consumePendingShiftClose") { () -> String? in
      ShiftLocationGuardManager.shared.consumePendingCloseReason()
    }
  }
}
